Azure Monitor Ingestion: Add troubleshooting guide (#34944)
Azure/azure-sdk-for-java: sdk/monitor/azure-monitor-ingestion/TROUBLESHOOTING.md

Environment variable name: `AZURE_HTTP_LOG_DETAIL_LEVEL`

| Value            | Logging level                                                        |
|------------------|----------------------------------------------------------------------|
| none             | HTTP request/response logging is disabled                            |
| basic            | Logs only URLs, HTTP methods, and time to finish the request.        |
| headers          | Logs everything in BASIC, plus all the request and response headers. |
| body             | Logs everything in BASIC, plus all the request and response body.    |
| body_and_headers | Logs everything in HEADERS and BODY.                                 |

**NOTE**: When logging the request and response bodies, ensure that they don't contain confidential information. When logging headers, the client library has a default set of headers that are considered safe to log. This set can be updated by updating the log options in the builder, as follows:

```java
clientBuilder.httpLogOptions(new HttpLogOptions().addAllowedHeaderName("safe-to-log-header-name"))
```

### Authentication errors

> Azure Monitor Ingestion supports Azure Active Directory authentication. The LogsIngestionClientBuilder can be configured to set the `credential`. To provide a valid credential, you can use `azure-identity` dependency. For more information on getting started, see the [README](https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/monitor/azure-monitor-ingestion#create-the-client) of the Azure Monitor Ingestion library. For more information on the credential types supported in `azure-identity`, see the [Azure Identity library documentation](https://learn.microsoft.com/azure/developer/java/sdk/identity).

Reviewer: ```suggestion
The Azure Monitor Ingestion library supports Azure Active Directory authentication. The `LogsIngestionClientBuilder` can be configured to set the `credential`. To provide a valid credential, you can use `azure-identity` dependency. For more information on getting started, see the [README](https://github.com/Azure/azure-sdk-for-java/tree/main/sdk/monitor/azure-monitor-ingestion#create-the-client) of the Azure Monitor Ingestion library. For more information on the credential types supported in `azure-identity`, see the [Azure Identity library documentation](https://learn.microsoft.com/azure/developer/java/sdk/identity).
```

# Troubleshooting Azure Monitor Ingestion client library issues

> This troubleshooting guide contains instructions to diagnose frequently encountered issues while using the Azure
> Monitor Ingestion client library for Java.

Reviewer: ```suggestion
This troubleshooting guide contains instructions to diagnose frequently encountered issues while using the Azure Monitor Ingestion client library for Java.
```

## Table of contents

> * [General Troubleshooting](#general-troubleshooting)

Reviewer: ```suggestion
* [General troubleshooting](#general-troubleshooting)
```

    * [Enable client logging](#enable-client-logging)
    * [Enable HTTP request/response logging](#enable-http-requestresponse-logging)
    * [Troubleshooting authentication issues](#authentication-errors)
    * [Troubleshooting NoSuchMethodError or NoClassDefFoundError](#dependency-conflicts)

> * [Troubleshooting Logs Ingestion](#troubleshooting-logs-ingestion)

Reviewer: ```suggestion
* [Troubleshooting logs ingestion](#troubleshooting-logs-ingestion)
```

    * [Troubleshooting authorization errors](#troubleshooting-authorization-errors)
    * [Troubleshooting missing logs](#troubleshooting-missing-logs)
    * [Troubleshooting slow logs upload](#troubleshooting-slow-logs-upload)

> ## General Troubleshooting

Reviewer: ```suggestion
## General troubleshooting
```

 ConfigurationId: <REDACTED>.."}}"
```

1. Check that the application or user that is making the request has sufficient permissions:
    * You can refer to this document
      to [manage access to data collection rule](https://learn.microsoft.com/azure/azure-monitor/logs/tutorial-logs-ingestion-portal#assign-permissions-to-the-dcr)
    * To ingest logs, ensure the service principal is assigned the "Monitoring Metrics Publisher" role for the data
      collection rule.
2. If the user or application is granted sufficient privileges to upload logs, make sure you are
   authenticating as that user/application. If you are authenticating using the
   [DefaultAzureCredential](https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/identity/azure-identity/README.md#authenticating-with-defaultazurecredential)
   then check the logs to verify that the credential used is the one you expected. To enable logging, see [enable
   client logging](#enable-client-logging) section above.
3. The permissions may take up to 30 minutes to propagate. So, if the permissions were granted recently, retry after
   some time.

### Troubleshooting missing logs

> When you send logs to Azure Monitor for ingestion, it's possible for the request to succeed, but you may not see the
> data appearing in the designated Log Analytics workspace table as configured in the Data Collection Rule. To investigate
> and resolve this issue, please ensure the following:

Reviewer: ```suggestion
When you send logs to Azure Monitor for ingestion, the request may succeed, but you may not see the data appearing in the designated Log Analytics workspace table as configured in the DCR. To investigate and resolve this issue, ensure the following:
```

- Double-check that you are using the correct data collection endpoint when configuring
  the `LogsIngestionClientBuilder`.
  Using the wrong endpoint can result in data not being properly sent to the Log Analytics workspace.

- Make sure you provide the correct data collection rule id (DCR ID) to the `upload` method. The DCR ID is an immutable
  identifier that determines the transformation rules applied to the uploaded logs and directs them to the appropriate
  Log Analytics workspace table.

> - Verify that the custom table specified in the data collection rule exists in the Log Analytics workspace. Ensure that
>   you provide the accurate name of the custom table to the upload method. Mismatched table names can lead to logs not
>   being stored correctly.
> 
> - Confirm that the logs you're sending adhere to the format expected by the data collection rule. The data should be in
>   the form of a JSON object or array, structured according to the requirements specified in the DCR. Additionally, it's
>   essential to encode the request body in UTF-8 to avoid any data transmission issues.
> 
> - Keep in mind that data ingestion may take some time, especially if you're sending data to a specific table for the

Reviewer: ```suggestion
* Verify that the custom table specified in the DCR exists in the Log Analytics workspace. Ensure that you provide the accurate name of the custom table to the upload method. Mismatched table names can lead to logs not being stored correctly.

* Confirm that the logs you're sending adhere to the format expected by the DCR. The data should be in the form of a JSON object or array, structured according to the requirements specified in the DCR. Additionally, it's essential to encode the request body in UTF-8 to avoid any data transmission issues.

* Keep in mind that data ingestion may take some time, especially if you're sending data to a specific table for the
```
